source/common/upstream/original_dst_cluster.h

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <string>
+
+#include "envoy/thread_local/thread_local.h"
+
+#include "common/common/empty_string.h"
+#include "common/common/logger.h"
+#include "common/upstream/upstream_impl.h"
+
+namespace Envoy {
+namespace Upstream {
+
+/**
+ * The OriginalDstCluster is a type of cluster that creates a single
+ * logical host that allows upstream connections to be opened to
+ * arbitrary IP destinations, as directed by metadata made available
+ * earlier in the filter pipeline.  Connection requests to the logical
+ * host itself will be denied.
+ *
+ * A new 'real host' is created for each destination, but currently
+ * stats and other state refers to the logical host.
+ */
+class OriginalDstCluster : public ClusterImplBase {
```

Comment:
Thanks, exactly the kind of feedback I was hoping for :-)

> The major problem I see with this approach (which is not quite the same as LogicalDnsCluster) is that you have the potential to add an unlimited # of hosts (and by association connection pools) that are never going to get cleaned up once they are created. At minimum I think we would need stats specific to this cluster type telling us how many real hosts we have. We don't specifically track the # of connection pools in stats for HTTP stuff, so it's going to be hard to tell exactly how many hosts/conn_pools we have created.

I had assumed that dynamically created hosts would get cleared when last reference to them dies out, and as they are not currently held in any containers that should happen soon after the connection is terminated? In testing I noticed that connection get's terminated right after the response is received.

> With that said, you can of course run out of memory in any number of other ways, and presumably you are using this cluster type in a specific deployment where you probably have control over what is happening, but I still feel that there should be some type of output telling us what is going on, even if the hosts aren't specifically in the primary cluster.

Assuming that we keep the dynamically created Hosts out of the cluster HostSets, do you have any suggestions on how to add stats like this?

> Note also that the way this is implemented your "real hosts" can/will be duplicated across all worker threads. I'm not sure if this matters or not.

Maybe I should make the logical host hold references to the created real hosts and reuse them both in the same thread and across the threads. This approach needs a timer to clean out unused hosts after a while, though.

> Stepping back a bit, there is another approach that could be taken (just sketching it out a bit):

> Write a filter that would sit inline to check to see if the original_dst is already in your cluster
If not, pause iteration, and do worker <-> main thread sync to add it to cluster, then continue iteration.

You mean inline with the http filters? And this is due to existing infra allowing filter iteration to be paused whereas no such mechanism exists at the cluster output stage?

How about making the router filter do the pausing if the cluster tells so, maybe via the new cluster->chooseHost() I added in this series?

> If it is, just continue, and do normal LB with some type of hint that picks a specific host in the load balancer by searching.

This might be best accomplished by a specialized load balancer that is designed to work like this. original_dst_cluster configuration could then be limited to that LB.

> I haven't thought through the previous very fully, but if that interests you, we can discuss further (and maybe should do a quick video conference).

At this stage I'd like learn more about the need to synchronize with the main thread when the host sets of the clusters are updated. I had a hunch it would be needed, and hence did not look deeper into that yet, hoping to come up with a solution avoiding such synchronization need, as it adds latency and complexity.

Right now I think maybe still hiding the real hosts behind one logical host, but making them (explicitly) stick around for some time to be reused and more explicitly managed with some kind of stats.

> I guess given how self contained this change is (mostly) I don't have a huge objection to the approach as long as the downsides are understood and well documented. LMK what you think.

My aim was to keep the changes to the overall architecture minimal. I'll look into adding more explicit management/stats of the real hosts and see if we can still avoid global synchronization for that. I'll wait for you responses to gauge which direction to go from here, though.

Thanks!